src/fabric.c

```diff
@@ -602,17 +602,30 @@ int DEFAULT_SYMVER_PRE(fi_getinfo)(uint32_t version, const char *node,
 		if (FI_VERSION_LT(prov->provider->fi_version, version))
 			continue;
 
-		if (ofi_is_util_prov(prov->provider) &&
-		    (flags & OFI_CORE_PROV_ONLY))
-			continue;
-
-		if (util_len && util_name) {
-			if (strncasecmp(util_name, prov->provider->name, util_len))
-				continue;
-		} else if (core_len && core_name) {
-			if (!ofi_is_util_prov(prov->provider) &&
-			    strncasecmp(core_name, prov->provider->name, core_len))
+		if (ofi_is_util_prov(prov->provider)) {
+			/* This is util provider */
+			if ((util_len && util_name) &&
+			    !(flags & OFI_CORE_PROV_ONLY)) {
+				/* If util name consists in provider name
+				 * and no OFI_CORE_PROV_ONLY flag is set,
+				 * FI info of util provider is requested */
+				if (strncasecmp(util_name,
+						prov->provider->name,
+						util_len))
+					continue;
+			} else {
+				/* Requested core provider's FI info 
+				 * or provider name doesn't contain util name */
```

Comment:
I would restructure the above code, so that the comments aren't as necessary.  If you first check for:
```
if (flags & OFI_CORE_PROV_ONLY)
    continue;
```
that check is straightforward.  Then check for util_len and util_name, and have it mirror the core_name check that you create below.  This will make it clear that the logic is the same, except for the handling of the flag.